### PR TITLE
librbd: retrieve image name when opening by id

### PR DESCRIPTION
**Pull request by trociny**
name is empty when the image is opened by its id.

Signed-off-by: Mykola Golub <mgolub@mirantis.com>

### CODE REVIEW

**Comment by dillaman** [comment]
@trociny Should we just fix the issue that name isn't populated when opening by ID?  We can add a new state to the image::RefreshRequest state machine to invoke the ```dir_get_name``` class method.

**Comment by trociny** [comment]
@dillaman

>  Should we just fix the issue that name isn't populated when opening by ID? We can add a new state to the image::RefreshRequest state machine to invoke the dir_get_name class method.

Ok, will try this way. Thanks.

**Comment by trociny** [comment]
@dillaman I have made the name to be populated when opening by ID.

Unfortunately it does not help for the initial issue, because asok is created earlier, on ImageCtx creation, before open. Do you know any hook I could use to create/delete asok on image open/close, instead of on image context construction/deletion?

**Comment by trociny** [comment]
@dillaman The same issue (empty name) is for perf name, which is initialized in ImageCtx::init().

**Comment by dillaman** [comment]
Actually, image::OpenRequest is probably a better state machine since we only need to load the name once.  ImageCtx::init should be invoked near the end of the refresh state machine, so as long as you load the name before you start the refresh, that should work.  As for the ASOK, can't we just delay it until ImageCtx::init?

**Comment by trociny** [comment]
@dillaman rebased after applying your suggestions.

File: src/librbd/image/OpenRequest.cc [hunk 1]
```diff
@@ -165,6 +165,42 @@ Context *OpenRequest<I>::handle_v2_get_id(int *result) {
...
+}
+
+template <typename I>
+Context *OpenRequest<I>::handle_v2_get_name(int *result) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 10) << __func__ << ": r=" << *result << dendl;
+
+  if (*result == 0) {
+    bufferlist::iterator it = m_out_bl.begin();
+    cls_client::dir_get_name_finish(&it, &m_image_ctx->name);
```
Comment (dillaman) [inline src/librbd/image/OpenRequest.cc]: *result = cls_client::dir_get_name_finish ...

File: src/librbd/image/OpenRequest.h [hunk 2]
```diff
@@ -38,7 +38,7 @@ class OpenRequest {
    *    \-----> V2_DETECT_HEADER                    |             .
    *                |                               |             .
    *                v                               |             .
-   *            V2_GET_ID                           |             .
+   *            V2_GET_ID|NAME                      |             .
```
Comment (dillaman) [inline src/librbd/image/OpenRequest.h]: To me, this implies that the _GET_ID state is mutually exclusive with the _GET_NAME state.

File: src/librbd/image/OpenRequest.cc (hunk 1 above)
Comment (trociny) [inline src/librbd/image/OpenRequest.cc]: Thanks! Fixed.

**Pull request merged**